PythonVisualizations/Stack.py

```diff
@@ -0,0 +1,322 @@
+        self.list = []
+        self.buttons = self.makeButtons()
+        self.maxArgWidth = maxArgWidth
+        self.display()
+
+    def __str__(self):
+        return str(self.list)
+
+    # Create an index arrow to point at an indexed cell with an optional name label
+    def createIndex(self, index, name=None):
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        x0 = self.STACK_X0 - self.CELL_WIDTH * 4 // 5
+        x1 = self.STACK_X0 - self.CELL_WIDTH * 3 // 10
+        y0 = y1 = cell_coords[1] + self.CELL_HEIGHT // 2
+
+        return self.drawArrow(
+            x0, y0, x1, y1, self.VARIABLE_COLOR, self.VARIABLE_FONT, name=("item", index))
+
+    # draw the actual arrow
+    def drawArrow(
+            self, x0, y0, x1, y1, color, font, name=None):  #
+        arrow = self.canvas.create_line(
+            x0, y0, x1, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x0 - self.CELL_WIDTH / 2, y0 + self.CELL_HEIGHT / 5, text=name, anchor=SW,
+                font=font, fill=color)
+
+        return (arrow, label) if name else (arrow,)
+
+    # STACK FUNCTIONALITY
+
+    def push(self, val):
+
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        # draw an index pointing to the last cell
+        indexDisplay = self.createIndex(len(self.list))
+        callEnviron |= set(indexDisplay)
+
+        cellCoords = self.cellCoords(len(self.list))
+        cellCenter = self.cellCenter(len(self.list))
+
+        # create new cell and cell value display objects
+        cellCoords = add_vector(cellCoords, (0, 0, 0, 2))
+        toPositions = (cellCoords, cellCenter)
+
+        # determine the top left and bottom right positions
+        startPosition = [self.STACK_X0, 0, self.STACK_X0, 0]
+        startPosition = add_vector(startPosition, (0, 0, self.CELL_WIDTH, self.CELL_HEIGHT))
+        cellPair = self.createCellValue(startPosition, val)
+        self.moveItemsTo(cellPair, toPositions, steps=self.CELL_HEIGHT, sleepTime=0.01)
+
+        # add a new Drawable with the new value, color, and display objects
+        d = drawable(
+            val, self.canvas.itemconfigure(cellPair[0], 'fill'), *cellPair)
+        self.list.append(d)
+
+        # finish the animation
+        self.cleanUp(callEnviron)
+
+    def pop(self):
+
+        # pop a Drawable from the list
+        if len(self.list) == 0:
+            return
+
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        n = self.list.pop()
+
+        # delete the associated display objects
+        items = (n.display_shape, n.display_val)
+
+        # self.moveItemsOffCanvas(items, N, sleepTime=0.02) ## alternative method to moveItemsBy
+        self.moveItemsBy(items, delta=(0, -400), steps=self.CELL_HEIGHT, sleepTime=.01)
+        self.canvas.delete(n.display_shape)
+        self.canvas.delete(n.display_val)
```

Comment:
While the `delta=(0, -400)` distance works OK with this canvas size, it would be better to ensure that the items move off the canvas regardless of their position at the time of the removal.  I would suggest using the lower y-coordinate of the display shape:
`delta=(0, -max(400, self.canvas.coords(n.display_shape)[3]))`
There's also the possibility of the user stopping the animation during the `moveItemsBy()` call.  That could leave the items on the screen and not cleaned up.  Since the items were already popped off the `self.list` in line 96, it would be safer to add them to the `callEnviron` and them take them back out after the calls to `self.canvas.delete()`.  Alternatively, we could leave them in the `callEnviron` and eliminate the calls to `self.canvas.delete()`, which would only delete them on the next cleanup operation.

Reply:
so is just deleting lines 105 and 106 good enough?

Reply:
Did you mean lines 103 and 104:
```
        self.canvas.delete(n.display_shape)
        self.canvas.delete(n.display_val)
```
Removing those two lines is almost good enough, but you need to add them to the `callEnviron` after they are created so that they are cleaned up later.  For example,
```
        callEnviron |= set(items)
```
after the `items` tuple is made in line 99.